### PR TITLE
EES-4667 Alter backend unit tests to use `ReleaseParent` where relevant

### PR DESCRIPTION
**Pull request by benoutram**
This PR alters backend unit tests to add in `ReleaseParent` relationships in test data where necessary for the planned switchover to use `ReleaseParent` to determine latest release versions.

In future, the `ReleaseRepository` methods added by https://github.com/dfe-analytical-services/explore-education-statistics/pull/4541 will make use of the new `ReleaseParent` hierarchy.

It's necessary for us to update all unit tests of any methods that indirectly rely on these and set up release versions with a parent entity.

Typically we currently see unit test data setup as follows with variations in the naming e.g. originalRelease/amendmentRelease , previousVersion/latestVersion etc...

```csharp
        var previousVersion = new Release
        {
            Id = Guid.NewGuid(),
            Publication = publication,
            Published = DateTime.UtcNow
        };

        var latestVersion = new Release
        {
            Id = Guid.NewGuid(),
            Publication = publication,
            PreviousVersionId = previousVersion.Id,
            ApprovalStatus = Draft
        };
```

`PreviousVersionId` was previously the only way to link related versions and is being used by the methods which determine latest release versions.

In future when we switch over to relying on `ReleaseParent`,  versions like in the example above need to be linked by a parent `ReleaseParent` entity instead.

This change depends on the `ReleaseParent` and `Publication` test data generator changes that were added by #4568 .

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleasePermissionServiceTests.cs [hunk 1]
```diff
@@ -527,26 +531,16 @@ await using (var contentDbContext = InMemoryApplicationDbContext(contentDbContex
...
-            };
-            var release2Amendment = new Release
-            {
-                Publication = publication,
-                ReleaseName = "2001",
-                TimePeriodCoverage = TimeIdentifier.AcademicYear,
-                PreviousVersion = release2Original,
-            };
+            ReleaseParent releaseParent1 = _dataFixture
+                .DefaultReleaseParent(publishedVersions: 1);
```
Comment (mmyoungman) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleasePermissionServiceTests.cs:535]: Just generally thinking about `DefaultReleaseParent`, I think with the arguments it currently takes, while it provides extreme convenience, could be potentially problematic due to the control you're giving up over the test data creation.

I imagine in the future, someone will want to create a parent with several releases that cannot be created with this method. And at that point they either alter the `releaseParent` after creating it via `_dataFixture`, or complicating the `_dataFixture` chain or maybe even having to override `DefaultReleaseParent` or make some other change such that they can customise the releases to get the test data they want.

Maybe that's fine. But I think it is something that needs to be managed carefully as I think this could turn messy quickly. I can imagine problems like lots of different ways of creating objects depending on the amount of customisation you require, or it being hard to follow what a particular `_dataFixture` chain is actually creating.

All that said, I could be overthinking this.

**Review by mmyoungman — APPROVED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs [hunk 2]
```diff
@@ -421,23 +421,23 @@ public class ReleaseService : IReleaseService
 
         public async Task<Either<ActionResult, DeleteDataFilePlan>> GetDeleteDataFilePlan(Guid releaseId, Guid fileId)
         {
-            return await _persistenceHelper.CheckEntityExists<Release>(releaseId)
+            return await _context.Releases
+                .FirstOrNotFoundAsync(r => r.Id == releaseId)
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs:425]: Yes we've got a comment about it on the backend style guide board to prefer "OrNotFound" extensions and phase out usage of `PersistenceHelper` because it adds no extra value yet requires setting up an extra dependency.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services.Tests/SubjectResultMetaServiceTests.cs [hunk 3]
```diff
@@ -279,13 +286,13 @@ public async Task GetSubjectMeta_BoundaryLevelViewModelsReturnedForSubject()
...
-                await contentDbContext.Publications.AddAsync(publication);
+                contentDbContext.Publications.Add(publication);
                 await contentDbContext.SaveChangesAsync();
             }
 
             await using (var statisticsDbContext = InMemoryStatisticsDbContext(statisticsDbContextId))
             {
-                await statisticsDbContext.ReleaseSubject.AddAsync(releaseSubject);
+                statisticsDbContext.ReleaseSubject.Add(releaseSubject);
                 await statisticsDbContext.SaveChangesAsync();
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services.Tests/SubjectResultMetaServiceTests.cs:296]: My interpretation of our recent decision to prefer `Add/AddRange` over `AddAsync/AddRangeAsync` is because the async methods add nothing.

The documentation for `AddAsync` and `AddRangeAsync` both state this remark:

>This method is async only to allow special value generators, such as the one used by 'Microsoft.EntityFrameworkCore.Metadata.SqlServerValueGenerationStrategy.SequenceHiLo', to access the database asynchronously. For all other cases the non async method should be used.

I don't think the same applies to `SaveChangesAsync` though, and I think generally we prefer to use async methods consistently over non-async methods, which is why I haven't changed these.

We could do with clarifying this decision with the rest of the team.

[the remark this answers is not in the document]
Comment (mmyoungman) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services.Tests/SubjectResultMetaServiceTests.cs:296]: I'd be surprised if it made any appreciable difference either way - the tests will run a fraction quicker or slower?

I'd be inclined to just forget I ever made a comment :)

[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services.Tests/SubjectResultMetaServiceTests.cs:296]: It's probably going to be negligible on the tests but the benefits of using async methods will apply to `SaveChangesAsync` which is why I've kept it in.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleasePermissionServiceTests.cs (hunk 1 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleasePermissionServiceTests.cs:535]: > Just generally thinking about DefaultReleaseParent, I think with the arguments it currently takes, while it provides extreme convenience, could be potentially problematic due to the control you're giving up over the test data creation.

I agree.

In some places that we still want to use test data generators for more complicated scenarios it might be better to split up the generation with declarations of different types into separate variables to make the composition a bit clearer, plus it can help avoid long expressions dereferencing particular objects e.g. `releases[0]` over `theme.topics[0].publications[0].releases[0]`.

The use of the new `DefaultReleaseParent(int publishedVersions, bool draftVersion = false, int? year = null)` isn't mandatory either. I just used it in pretty much every change on this PR for convenience.

It's still possible for example to do something like this and keep more control on what extra properties are set:

```csharp
            var releases = _dataFixture
                .DefaultRelease()
                .Generate(1);

            ReleaseParent releaseParent = _dataFixture
                .DefaultReleaseParent()
                .WithReleases(releases);
```

or

```csharp
            Release release = _dataFixture
                .DefaultRelease()
                .WithReleaseParent(_dataFixture
                    .DefaultReleaseParent());
```

> I imagine in the future, someone will want to create a parent with several releases that cannot be created with this method. And at that point they either alter the releaseParent after creating it via _dataFixture, or complicating the _dataFixture chain or maybe even having to override DefaultReleaseParent or make some other change such that they can customise the releases to get the test data they want.

I agree, I think they will. I haven't spent too much time thinking about what the signature of it should be and what the possible use cases of it are. I did seem to be able to retrofit it into pretty much all the unit tests I wanted to update as part of this PR though without too much complication.

>  I can imagine problems like lots of different ways of creating objects depending on the amount of customisation you require, or it being hard to follow what a particular _dataFixture chain is actually creating.

I agree, if there's a lot of customisation then it's probably best to declare separate variables with customisation first and compose a hierarchy from those if necessary.

We can use of `ToTuple2` and `ToTuple3` to dereference created objects with explicit variable naming to help avoid having to understand the configuration, e.g.

```csharp
            var (publication, unpublishedPublication) = _dataFixture
                .DefaultPublication()
                ...
                .Generate(2)
                .ToTuple2();
```

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services.Tests/SubjectResultMetaServiceTests.cs [hunk 4]
```diff
@@ -203,18 +204,24 @@ await using (var statisticsDbContext = InMemoryStatisticsDbContext(statisticsDbC
...
+
+            Subject subject = _dataFixture
+                .DefaultSubject();
+
+            ReleaseSubject releaseSubject = _dataFixture
+                .DefaultReleaseSubject()
+                .WithRelease(_dataFixture
+                    .DefaultStatsRelease()
+                    .WithId(release.Id)
+                    .WithPublicationId(publication.Id))
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services.Tests/SubjectResultMetaServiceTests.cs:223]: I think I'm going to leave this one if it's ok with you.

> A simple function call like CreateReleaseSubject(release, subject)

We'd still need to convert the `ContentModel.Release` to a `Data.Model.Release` either prior to calling that, or as part of it.

> If we want to use _dataFixture, maybe we can have WithRelease(release)? Or have DefaultReleaseSubject take arguments like DefaultReleaseSubject(release, subject)?

These have the same problem. I don't think we should add a dependency on `Content.Model` to `Data.Services.Tests` just for the convenience of being able to make the call with a `Content.Model.Release`.

There's also some examples in `TableBuilderServiceTests` where a subject is not passed, so left to be set up with a default subjectId, and an example of the stats release where the publication id is different from the content model's release publicationId.

**Review by mmyoungman — APPROVED** [review]
